libraries/Microsoft.Bot.Builder.Core.Extensions/TranscriptLoggerMiddleware.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.Core.Extensions
+{
+    /// <summary>
+    /// When added, this middleware will log incoming and outgoing activitites to a ITranscriptStore 
+    /// </summary>
+    public class TranscriptLoggerMiddleware : IMiddleware
+    {
+        private static JsonSerializerSettings JsonSettings = new JsonSerializerSettings() { NullValueHandling = NullValueHandling.Ignore };
+        private ITranscriptLogger logger;
+
+        private Queue<IActivity> transcript = new Queue<IActivity>();
+
+        /// <summary>
+        /// Middleware for logging incoming and outgoing activities to a transcript store 
+        /// </summary>
+        /// <param name="transcriptLogger"></param>
+        public TranscriptLoggerMiddleware(ITranscriptLogger transcriptLogger)
+        {
+            logger = transcriptLogger ?? throw new ArgumentNullException("TranscriptLoggerMiddleware requires a ITranscriptLogger implementation.  ");
+        }
+
+        /// <summary>
+        /// initialization for middleware turn
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="nextTurn"></param>
+        /// <returns></returns>
+        public async Task OnTurn(ITurnContext context, MiddlewareSet.NextDelegate nextTurn)
+        {
+            // log incoming activity at beginning of turn
+            if (context.Activity != null)
+            {
+                if (String.IsNullOrEmpty((String)context.Activity.From.Properties["role"]))
+                    context.Activity.From.Properties["role"] = "user";
+
+                LogActivity(CloneActivity(context.Activity));
+            }
+
+            // hook up onSend pipeline
+            context.OnSendActivities(async (ctx, activities, nextSend) =>
+            {
+                // run full pipeline
+                var responses = await nextSend();
+
+                foreach (var activity in activities)
+                {
+                    LogActivity(CloneActivity(activity));
```

Comment:
>CloneActivity [](start = 32, length = 13)

You do this in several places. Wouldn't it make the code more robust to just have LogActivity() always clone first?

Reply:
Oh I see there are places where you call it without clone.

---
In reply to: [181250465](https://github.com/Microsoft/botbuilder-dotnet/pull/410#discussion_r181250465) [](ancestors = 181250465)